File: ts/test/fixture/transport/context.rb
require_relative 'core/context'

root = DemoContext.new({ 'config' => {} })
entity = DemoContext.new({}, root)
first = DemoContext.new({ 'opname' => 'list' }, entity)
first.ctrl.paging = { 'cursor' => 'first' }
second = DemoContext.new({ 'opname' => 'list' }, entity)
raise 'shared operation control' if first.ctrl.equal?(second.ctrl)
raise 'root paging changed' unless root.ctrl.paging.nil?
raise 'paging leaked' unless second.ctrl.paging.nil?
raise 'nested control lost' unless DemoContext.new({}, first).ctrl.equal?(first.ctrl)
paging = { 'cursor' => 'explicit' }
explicit = DemoContext.new({ 'opname' => 'list', 'ctrl' => { 'paging' => paging } }, entity)
raise 'explicit paging lost' unless explicit.ctrl.paging.equal?(paging)
puts 'context: isolated defaults; explicit and nested controls preserved'

File: ts/test/fixture/transport/context.pl
use strict;
use warnings;
use Scalar::Util qw(refaddr);
require './core/context.pm';

my $root = DemoContext->new({config => {}});
my $entity = DemoContext->new({}, $root);
my $first = DemoContext->new({opname => 'list'}, $entity);
$first->{ctrl}{paging} = {cursor => 'first'};
my $second = DemoContext->new({opname => 'list'}, $entity);
die 'shared operation control' if refaddr($first->{ctrl}) == refaddr($second->{ctrl});
die 'root paging changed' if defined $root->{ctrl}{paging};
die 'paging leaked' if defined $second->{ctrl}{paging};
die 'nested control lost' unless refaddr(DemoContext->new({}, $first)->{ctrl}) == refaddr($first->{ctrl});
my $paging = {cursor => 'explicit'};
my $explicit = DemoContext->new({opname => 'list', ctrl => {paging => $paging}}, $entity);
die 'explicit paging lost' unless refaddr($explicit->{ctrl}{paging}) == refaddr($paging);
print "context: isolated defaults; explicit and nested controls preserved\n";
